src/post_transform.c

```diff
@@ -262,32 +260,263 @@ int pluto_intra_tile_optimize_band(Band *band, int num_tiled_levels,
+                                    PlutoProg *prog) {
+  /* Band has to be the innermost band as well */
+  if (!pluto_is_band_innermost(band, num_tiled_levels, 0)) {
+    /* If the loop nest is tiled (but not completely), then move an intratile
+     * parallel loop in this band to the innermost level. */
     if (num_tiled_levels > 0) {
-      move_across_scalar_hyperplanes = true;
+      return make_parallel_loop_innermost(band, num_tiled_levels, prog);
     }
-    pluto_make_innermost_loop(best_loop, last_level,
-                              move_across_scalar_hyperplanes, prog);
-    pluto_loops_free(loops, nloops);
-    return 1;
+    return false;
   }
 
-  pluto_loops_free(loops, nloops);
-  return 0;
+  /* Band may have been distributed in the inter tile space.  If the band is
+   * distributed then there can be multiple innner permutable bands. We need to
+   * find optimize for each of them separately. First find bands with a single
+   * statement in it and then fuse the bands with statements that are fused in
+   * the tile space. */
+  /* TODO: We need an early bailout condition here. If the number of statements
+   * in the band is 1 or the loop nest is not tiled, then we can skip the
+   * procedure to find inner most bands corresponding to the input band. */
+  unsigned nstmt_bands = 0;
+  Band **per_stmt_bands = get_per_stmt_band(band, &nstmt_bands);
+
+  unsigned num_fused_bands = 0;
+  Band **ibands =
+      fuse_per_stmt_bands(per_stmt_bands, nstmt_bands, num_tiled_levels,
+                          &num_fused_bands, prog->context);
+
+  if (num_fused_bands != nstmt_bands) {
+    pluto_bands_free(per_stmt_bands, nstmt_bands);
+  }
+  PlutoContext *context = prog->context;
+  PlutoOptions *options = context->options;
+  if (options->debug) {
+    printf("Bands for intra tile optimiztion \n");
+    pluto_bands_print(ibands, num_fused_bands);
+  }
+
+  bool retval = false;
+  for (unsigned i = 0; i < num_fused_bands; i++) {
+    Band *band = ibands[i];
+    int depth = band->loop->depth + num_tiled_levels * band->width;
+    IF_DEBUG(printf("Getting loop at depth  %d\n", depth););
+
+    unsigned nloops;
+    Ploop **loops = pluto_get_loops_under(band->loop->stmts, band->loop->nstmts,
+                                          depth, prog, &nloops);
+    Ploop *best_loop = get_best_vectorizable_loop(loops, nloops, prog);
+
+    if (best_loop && !pluto_loop_is_innermost(best_loop, prog)) {
+      IF_DEBUG(printf("[pluto] intra_tile_opt: loop to be made innermost: "););
+      IF_DEBUG(pluto_loop_print(best_loop););
+
+      /* The last level in the innermost permutable band. This is true only if
+       * the outermost permutable band and innermost permutable band are the
+       * same. */
+      unsigned last_level =
+          band->loop->depth + (num_tiled_levels + 1) * band->width;
+      bool move_across_scalar_hyperplanes = false;
+      /* Move loop across scalar hyperplanes only if the loop nest is tiled.*/
```

Comment:
I can't recall what the reasoning behind this was or if there was one. We should document that as well or just put a TODO/FIXME for further exploration.